### PR TITLE
Refactored asset search by identifiers

### PR DESCRIPTION
**Pull request by vanruch**

### CODE REVIEW

File: src/services/data_model_engine.js [hunk 1]
```diff
@@ -103,10 +104,25 @@ export default class DataModelEngine {
...
+    return [...new Set(events.results.map((event) => event.content.idData.assetId))];
+  }
+
   async findAssets(params, tokenData) {
-    const validatedParams = this.entityBuilder.validateAndCastFindAssetsParams(params);
-    const accessLevel = await this.accountAccessDefinitions.getTokenCreatorAccessLevel(tokenData);
-    const findAssetQueryObject = this.findAssetQueryObjectFactory.create(validatedParams, accessLevel);
+    let validatedParams = this.entityBuilder.validateAndCastFindAssetsParams(params);
+    if (validatedParams.identifier) {
+      const accessLevel = await this.accountAccessDefinitions.getTokenCreatorAccessLevel(tokenData);
```
Comment (Anze1m) [inline src/services/data_model_engine.js]: Why is this line inside of an `if`?

Comment (Anze1m) [inline src/services/data_model_engine.js]: Why is this line inside of an `if`?

File: src/services/data_model_engine.js [hunk 2]
```diff
@@ -103,10 +104,25 @@ export default class DataModelEngine {
...
+      const consideredAssets = await this.selectAssetsByIdentifier(validatedParams.identifier, accessLevel);
+      validatedParams = pick(put(validatedParams, 'consideredAssets', consideredAssets), 'identifier');
```
Comment (spherefoundry) [inline src/services/data_model_engine.js]: I would change `consideredAssets` to just `assetIds`. In the future we could then introduce direct querying of multiple assets by theirs assetsIds.

File: src/services/data_model_engine.js [hunk 3]
```diff
@@ -103,10 +104,25 @@ export default class DataModelEngine {
     return asset;
   }
 
+  async selectAssetsByIdentifier(identifiers, accessLevel) {
```
Comment (spherefoundry) [inline src/services/data_model_engine.js]: `identifiers` => `identifier` as this may be misunderstood as taking multiple identifiers

Comment (spherefoundry) [reply src/services/data_model_engine.js]: Also the name should be changed to reflect we are returning AssetIDS and not assets directly.

File: test/services/data_model_engine.js [hunk 4]
```diff
@@ -515,6 +503,85 @@ describe('Data Model Engine', () => {
...
+            {content: {idData: {assetId: '10', one: {two: 2}}}},
+            {content: {idData: {assetId: '1'}}}
+          ]
+        });
+        mockAccountAccessDefinitions.getTokenCreatorAccessLevel.resolves(accessLevel);
+
+        modelEngine = new DataModelEngine({}, {}, mockEntityBuilder, {},
+          {}, {}, {}, mockFindEventQueryObjectFactory, {}, mockFindAssetQueryObjectFactory, mockAccountAccessDefinitions);
+
+        result = await modelEngine.selectAssetsByIdentifier(identifier, accessLevel);
```
Comment (spherefoundry) [inline test/services/data_model_engine.js]: you are only using this inside of one it case, so this should move;

File: test/services/data_model_engine.js [hunk 5]
```diff
@@ -515,6 +503,85 @@ describe('Data Model Engine', () => {
...
+        }, accessLevel);
+      });
+
+      it('should execute and return unique ids', () => {
+        expect(mockFindEventQueryObject.execute).to.be.calledOnce;
+        expect(result).to.deep.equal(['1', '10']);
+      });
+
+      it('when accessLevel was not given it defaults to 0', async () => {
+        await modelEngine.selectAssetsByIdentifier(identifier);
```
Comment (spherefoundry) [inline test/services/data_model_engine.js]: Tests for different subjects should be grouped at  least into separate describe blocks

File: test/services/find_event_query_object.js [hunk 6]
```diff
@@ -272,6 +281,51 @@ describe('Find Event Query Object', () => {
...
+
+        it('finds events when select by several ids', async () => {
+          findEventQueryObject = findEventQueryObjectFactory.create(
+            makeAssetIdentifierSearchParams({id: '2', id2: 'abc'}), 100);
+          const found = await findEventQueryObject.execute();
+          expect(found.results).to.deep.equal([eventsSet[2]]);
+          expect(found.resultCount).to.equal(1);
+        });
+
+        it('omits events when the access level in event was higher than user\'s', async () => {
```
Comment (spherefoundry) [inline test/services/find_event_query_object.js]: you can skip the `\s` by using \` instead of \' for the whole string

**Pull request merged**